### PR TITLE
Add StorageIceberg and table function iceberg

### PR DESCRIPTION
**Pull request by ucasfl**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `StorageIceberg` and table function `iceberg` to access iceberg table store on S3.

### CODE REVIEW

File: src/Storages/StorageIceberg.cpp [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+    StorageS3::updateS3Configuration(ctx, base_configuration);
+    auto new_configuration = getAdjustedS3Configuration(
+        ctx, base_configuration, configuration, base_configuration.uri.key, &Poco::Logger::get("StorageIceberg"));
+    return StorageS3::getTableStructureFromData(
+        new_configuration, /*distributed processing*/ false, format_settings, ctx, /*object_infos*/ nullptr);
+}
+
+void registerStorageIceberg(StorageFactory & factory)
+{
+    factory.registerStorage(
```
Comment (stigsb) [inline src/Storages/StorageIceberg.cpp:386]: Could you add support for "named collections" here? Check src/Storages/ExternalDataSourceConfiguration.h

Comment (kssenii) [reply src/Storages/StorageIceberg.cpp:386]: > Check src/Storages/ExternalDataSourceConfiguration.h

this is outdated code, which is almost removed, check `src/Common/NamedCollections.h` instead. And `src/Storages/NamedCollectionsHelpers.h`.

File: src/TableFunctions/TableFunctionIceberg.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+
+void TableFunctionIceberg::parseArgumentsImpl(
```
Comment (kssenii) [inline src/TableFunctions/TableFunctionIceberg.cpp:30]: let's reuse code from s3 table function as it is the same here, for example you can inherit from it.

File: src/Storages/StorageIceberg.cpp [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+{
+    /// Iceberg stores all the metadata.json in metadata directory, and the
+    /// newest version has the max version name, so we should list all of them
+    /// then find the newest metadata.
+    std::vector<String> metadata_files;
+
+    const auto & client = base_configuration.client;
+
+    Aws::S3::Model::ListObjectsV2Request request;
+    Aws::S3::Model::ListObjectsV2Outcome outcome;
```
Comment (kssenii) [inline src/Storages/StorageIceberg.cpp:79]: let's make a function in s3Common like `listFiles` by some prefix and reuse it here

File: src/Storages/StorageIceberg.cpp [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+        }
+    }
+
+    return keys;
+}
+
+std::shared_ptr<ReadBuffer> IcebergMetaParser::createS3ReadBuffer(const String & key) const
+{
+    S3Settings::RequestSettings request_settings;
+    request_settings.max_single_read_retries = 10;
```
Comment (kssenii) [inline src/Storages/StorageIceberg.cpp:260]: why hardcode like this? may be use value from setting?

File: src/TableFunctions/TableFunctionIceberg.cpp (hunk 2 above)
Comment (ucasfl) [reply src/TableFunctions/TableFunctionIceberg.cpp:30]: It's not totally same here, If we don't specify format for it, we can not get format from it and should use `Parquet` as default file format.

File: src/Storages/StorageIceberg.cpp (hunk 1 above)
Comment (ucasfl) [reply src/Storages/StorageIceberg.cpp:386]: I don' t understand what here mean.

Comment (kssenii) [reply src/Storages/StorageIceberg.cpp:353]: https://clickhouse.com/docs/en/operations/named-collections/

File: src/TableFunctions/ITableFunctionDataLake.h [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+public:
+    static constexpr auto name = Name::name;
+    std::string getName() const override
+    {
+        return name;
+    }
+
+protected:
+    StoragePtr
+    executeImpl(const ASTPtr & /*ast_function*/, ContextPtr context, const std::string & table_name, ColumnsDescription /*cached_columns*/)
```
Comment (kssenii) [inline src/TableFunctions/ITableFunctionDataLake.h:44]: let's move this code into cpp and add something like this https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/NamedCollections/NamedCollections.cpp#L435 to resolve the templates.

File: src/Storages/StorageIceberg.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+// Class to parse iceberg metadata and find files needed for query in table
+// Iceberg table directory outlook:
+// table/
+//      data/
+//      metadata/
+// The metadata has three layers: metadata -> manifest list -> manifest files
+class IcebergMetaParser
+{
+public:
+    IcebergMetaParser(const StorageS3::S3Configuration & configuration_, const String & table_path_, ContextPtr context_);
```
Comment (kssenii) [inline src/Storages/StorageIceberg.h:34]: Here it is better to get rid of s3 too and make it abstract. May be use reference to a base class I added...
https://github.com/ClickHouse/ClickHouse/blob/ab0dedf0c88cbfb732d7c0c71922dce49a227607/src/Storages/StorageConfiguration.h ? (Merge this branch into your PR?)

File: src/TableFunctions/TableFunctionS3.h [hunk 7]
```diff
@@ -37,8 +36,10 @@ class TableFunctionS3 : public ITableFunction
         return {"_path", "_file"};
     }
 
+    template <bool get_format_from_file = true>
+    static void parseArgumentsImpl(const String & error_message, ASTs & args, ContextPtr context, StorageS3::Configuration & configuration);
```
Comment (kssenii) [inline src/TableFunctions/TableFunctionS3.h:40]: why not make `get_format_from_file` as just an argument? this template adds extra complexity in code for no reason, this is usually used for performance, but in our case it will not make any difference.

File: src/Storages/StorageS3.h [hunk 8]
```diff
@@ -35,6 +35,23 @@ class PullingPipelineExecutor;
...
+
+struct S3DataLakeMetaReadHelper;
+
+struct StorageIcebergName;
+template <typename Configuration, typename MetaReadHelper>
+class IcebergMetaParser;
+
+struct StorageDeltaLakeName;
+template <typename Configuration, typename MetaReadHelper>
+class DeltaLakeMetaParser;
```
Comment (kssenii) [inline src/Storages/StorageS3.h:49]: better to use full word `Metadata` instead of `Meta`

File: src/Storages/IStorageDataLake.h [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    static StorageS3::Configuration
+    getAdjustedS3Configuration(const ContextPtr & context, StorageS3::Configuration & configuration, Poco::Logger * log)
+    {
+        MetaParser parser{configuration, context};
+
+        auto keys = parser.getFiles();
+        String new_uri = std::filesystem::path(configuration.url.uri.toString()) / Name::data_directory_prefix
+            / MetaParser::generateQueryFromKeys(keys, configuration.format);
```
Comment (kssenii) [inline src/Storages/IStorageDataLake.h:148]:
```suggestion
            / MetadataParser::generateQueryFromKeys(keys, configuration.format);
```

File: src/Storages/S3DataLakeMetaReadHelper.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    request_settings.max_single_read_retries = context->getSettingsRef().s3_max_single_read_retries;
+    return std::make_shared<ReadBufferFromS3>(
+        base_configuration.client,
+        base_configuration.url.bucket,
+        key,
+        base_configuration.url.version_id,
+        request_settings,
+        context->getReadSettings());
+}
+std::vector<String> S3DataLakeMetaReadHelper::listFilesMatchSuffix(
```
Comment (kssenii) [inline src/Storages/S3DataLakeMetaReadHelper.cpp:34]:
```suggestion
}

std::vector<String> S3DataLakeMetaReadHelper::listFilesMatchSuffix(
```

File: src/Storages/S3DataLakeMetaReadHelper.h [hunk 11]
```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <config.h>
+
+#if USE_AWS_S3
+
+#    include <IO/ReadBufferFromS3.h>
```
Comment (kssenii) [inline src/Storages/S3DataLakeMetaReadHelper.h:7]: this include seems redundant

File: src/Storages/S3DataLakeMetaReadHelper.h [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+#    include <IO/ReadHelpers.h>
+#    include <Storages/StorageS3.h>
+
+#    include <Formats/FormatFactory.h>
```
Comment (kssenii) [inline src/Storages/S3DataLakeMetaReadHelper.h:11]: this one either

File: src/Storages/StorageDeltaLake.cpp [hunk 13]
```diff
@@ -57,20 +56,22 @@ std::vector<String> DeltaLakeMetadata::listCurrentFiles() &&
     return keys;
 }
 
-JsonMetadataGetter::JsonMetadataGetter(const StorageS3::Configuration & configuration_, ContextPtr context)
+template <typename Configuration, typename MetaReadHelper>
```
Comment (kssenii) [inline src/Storages/StorageDeltaLake.cpp:59]: I'd prefer it was done with inheritance instead of templates, but templates are ok, no need to change. But when remote storages are rewritten to use the interface of IObjectStorage (from Disks/IObjectStorage) this code will probably rewritten. But not soon, so ok for now with templates.

File: src/Storages/IStorageDataLake.h [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+    static ColumnsDescription getTableStructureFromData(
+        StorageS3::Configuration & configuration, const std::optional<FormatSettings> & format_settings, ContextPtr ctx)
+    {
+        StorageS3::updateS3Configuration(ctx, configuration);
+        auto new_configuration = getAdjustedS3Configuration(ctx, configuration, &Poco::Logger::get("Storage" + String(name)));
+        return StorageS3::getTableStructureFromData(
+            new_configuration, /*distributed processing*/ false, format_settings, ctx, /*object_infos*/ nullptr);
+    }
+
+    static StorageS3::Configuration
```
Comment (kssenii) [inline src/Storages/IStorageDataLake.h:141]: not sure what was the point in all those templates if we anyway use s3 explicitly?

File: src/Storages/IStorageDataLake.h [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+
+        StorageS3::Configuration new_configuration(configuration);
+        new_configuration.url = S3::URI(new_uri);
+
+        LOG_DEBUG(log, "Table path: {}, new uri: {}", configuration.url.key, new_uri);
+
+        return new_configuration;
+    }
+
+    static void processNamedCollectionResult(StorageS3::Configuration & configuration, const NamedCollection & collection)
```
Comment (kssenii) [inline src/Storages/IStorageDataLake.h:158]: this method is the same as in StorageS3? Why not reuse then?

File: src/Storages/IStorageDataLake.h (hunk 14 above)
Comment (ucasfl) [reply src/Storages/IStorageDataLake.h:141]: I get rid of S3 related things in above metadata parser, such we can easily to implement parser if we want to read the data lake tables store on other filesystem like HDFS. Buf for `IStorageDataLake` class, it's not easy to get rid of S3 related things.

File: src/Storages/IStorageDataLake.h (hunk 15 above)
Comment (ucasfl) [reply src/Storages/IStorageDataLake.h:158]: It's difference, in StorageS3, it may specify a file name, but this is not.

Comment (kssenii) [reply src/Storages/IStorageDataLake.h:137]: not possible - is it because we need to parse arguements in a certain way different from for example StorageS3? E.g. we do not need `compression_method` and `format` arguments as the first one is not supported and the second one should be Parquet always? Then seems like we can allow to parse the arguments as StorageS3 usually does, and then check the result having just a base class of `S3Storage::Configuration : public StatelessTableEngineConfiguration` (from Storages/StorageConfiguration.h) which has format and compression methods, and throw exception or rewrite them if needed. WDYT? E.g. as I understand it is enough for us to have pointer to base StatelessTableEngineConfiguration here to resolve the differences.

**Comment by didip** [comment]
What an amazing effort! I hope that this will be merged soon!

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 16]
```diff
@@ -68,35 +68,35 @@ namespace ErrorCodes
     extern const int CANNOT_READ_ALL_DATA;
 }
 
-class InputStreamReadBufferAdapter : public avro::InputStream
+bool AvroInputStreamReadBufferAdapter::next(const uint8_t ** data, size_t * len)
```
Comment (kssenii) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp:71]: what was the reason for these changes in Avro? May be it can be submitted in a separate PR?

File: src/Storages/IStorageDataLake.h [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+        const StorageS3::Configuration & configuration_,
+        const StorageID & table_id_,
+        ColumnsDescription columns_,
+        const ConstraintsDescription & constraints_,
+        const String & comment,
+        ContextPtr context_,
+        std::optional<FormatSettings> format_settings_)
+        : IStorage(table_id_)
+        , base_configuration{configuration_}
+        , log(&Poco::Logger::get("Storage" + String(name) + "(" + table_id_.table_name + ")"))
```
Comment (kssenii) [inline src/Storages/IStorageDataLake.h:36]: too complex, let's use `fmt::format`

File: src/Storages/StorageIceberg.cpp [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+    /// newest version has the max version name, so we should list all of them
+    /// then find the newest metadata.
+    static constexpr auto meta_file_suffix = ".json";
+    auto metadata_files = MetadataReadHelper::listFilesMatchSuffix(base_configuration, metadata_directory, meta_file_suffix);
+
+    if (metadata_files.empty())
+        throw Exception(
+            ErrorCodes::FILE_DOESNT_EXIST, "The metadata file for Iceberg table with path {} doesn't exist", base_configuration.url.key);
+
+    auto it = std::max_element(metadata_files.begin(), metadata_files.end());
```
Comment (kssenii) [inline src/Storages/StorageIceberg.cpp:68]: why do we need to take max?

File: src/TableFunctions/ITableFunctionDataLake.h [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+            " - url, access_key_id, secret_access_key, format, structure\n"
+            " - url, access_key_id, secret_access_key, format, structure, compression_method",
+            getName());
+
+        if (args_func.size() != 1)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Table function '{}' must have arguments", getName());
+
+        auto & args = args_func.at(0)->children;
+
+        TableFunctionS3::parseArgumentsImpl(message, args, context, configuration, false);
```
[the remark this answers is not in the document]
Comment (ucasfl) [reply src/TableFunctions/ITableFunctionDataLake.h:83]: OK.

File: src/TableFunctions/ITableFunctionDataLake.h [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+        return parseColumnsListFromString(configuration.structure, context);
+    }
+
+    void parseArguments(const ASTPtr & ast_function, ContextPtr context) override
+    {
+        /// Parse args
+        ASTs & args_func = ast_function->children;
+
+        const auto message = fmt::format(
```
[the remark this answers is not in the document]
Comment (ucasfl) [reply src/TableFunctions/ITableFunctionDataLake.h:66]: The message?

File: src/Storages/StorageS3.h [hunk 21]
```diff
@@ -302,8 +318,9 @@ class StorageS3 : public IStorage, WithContext
 private:
     friend class StorageS3Cluster;
     friend class TableFunctionS3Cluster;
-    friend class StorageHudi;
-    friend class StorageDeltaLake;
+    friend class IStorageDataLake<StorageHudiName, HudiMetadataParser<StorageS3::Configuration, S3DataLakeMetadataReadHelper>>;
```
[the remark this answers is not in the document]
Comment (ucasfl) [reply src/Storages/StorageS3.h:321]: Not needed anymore. The method has been moved to public.

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp (hunk 16 above)
Comment (ucasfl) [reply src/Processors/Formats/Impl/AvroRowInputFormat.cpp:71]: I used some method here to parse the avro format iceberg meta file.

File: src/Storages/StorageIceberg.cpp (hunk 18 above)
Comment (ucasfl) [reply src/Storages/StorageIceberg.cpp:68]: The iceberg table is multi version, each version corresponding to a `.json` metadata file, and the newest version has the largest version number.

Comment (kssenii) [reply src/Storages/StorageIceberg.cpp:68]: ok, let's put this explanation into comment in code?

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp (hunk 16 above)
Comment (kssenii) [reply src/Processors/Formats/Impl/AvroRowInputFormat.cpp:71]: Ok

File: src/TableFunctions/ITableFunctionDataLake.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (ucasfl) [reply src/TableFunctions/ITableFunctionDataLake.h:66]: The `message`?

File: src/Storages/StorageS3.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (kssenii) [reply src/Storages/StorageS3.h:321]: btw, I now had a quick idea/question : why don't we inherit IStorageDeltaLake from template Storage (S3Storage), would it be more clean and readable? I am not sure at the moment whether it will be better, but at least then there will be not need in moving methods to `public`, they could be `protected`

[the remark this answers is not in the document]
Comment (kssenii) [reply src/Storages/StorageS3.h:321]: And also it will allow to get rid of template configuration structure, as we just will be able to use `Storage::Configuration` instead.

File: src/Storages/StorageIceberg.cpp (hunk 18 above)
Comment (ucasfl) [reply src/Storages/StorageIceberg.cpp:68]: I have put comment in the beginning of the method.

File: src/Storages/StorageS3.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (ucasfl) [reply src/Storages/StorageS3.h:321]: Maybe it's not a good idea. For example, the IStorageDataLake does not support `write` and `truncate`, also does not have virtual column, so we also need to override these related methods.

File: src/TableFunctions/ITableFunctionDataLake.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (kssenii) [reply src/TableFunctions/ITableFunctionDataLake.h:66]: yeah, this will allow to make it not dependant on s3, but also comment from https://github.com/ClickHouse/ClickHouse/pull/45384#discussion_r1108375320 applies here too. It will allow to get rid of multiple templates and will require only one Storage template.

File: src/Storages/StorageS3.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (kssenii) [reply src/Storages/StorageS3.h:321]: > For example, the IStorageDataLake does not support write and truncate, also does not have virtual column, so we also need to override these related methods.

override and throw exception "Not supported"? It seems better than having a lot of templates for everything: for Storage, for MetadataReadHelper, for Configuration. If we inherit from Storage, then we can get Storage::Configuration and everything else needed from storage. You don't like the idea still? Just now we still have s3 mentions in StorageDeltaLake, for which you mentioned it is difficult to get rid of. So inheriting from needed storage will allow to get rid of s3 in that case?

[the remark this answers is not in the document]
Comment (ucasfl) [reply src/Storages/StorageS3.h:321]: Currently, the class just has two templates: `Name` and `MetadataParser`. I think is OK to inherit from the Storage.

[the remark this answers is not in the document]
Comment (ucasfl) [reply src/Storages/StorageS3.h:321]: But, if inherit from StorageS3, how can I initialize the base class, we need first parse the metadata and update configuration, so current use PImpl.

[the remark this answers is not in the document]
Comment (kssenii) [reply src/Storages/StorageS3.h]: hm, ok.
But let's just use Storage::Configuration instead of template Configuration, because definition of configuration will be visible even with uninitialised storage.

[the remark this answers is not in the document]
Comment (kssenii) [reply src/Storages/StorageS3.h]: > But, if inherit from StorageS3, how can I initialize the base class, we need first parse the metadata and update configuration, so current use PImpl.

This issue will be resolved if you use variadic argument for constructor of IStorageDataLake.

[the remark this answers is not in the document]
Comment (ucasfl) [reply src/Storages/StorageS3.h]: I don't understand how to resolve if by variadic arguments. We still need to initialize StorageS3 in initialization list.

[the remark this answers is not in the document]
Comment (kssenii) [reply src/Storages/StorageS3.h]: something like:
```
template <class Storage, class ...Args>
explicit IStorageDataLake(Args && ...args) : Storage(std::forward<Args>(args)...) {}
```

[the remark this answers is not in the document]
Comment (ucasfl) [reply src/Storages/StorageS3.h]: No, it's not about pass arguments, its becuase the arguments of IStorageDataLake different with StorageS3.

[the remark this answers is not in the document]
Comment (kssenii) [reply src/Storages/StorageS3.h]: Currently I see they are the same except for `getAdjustedConfiguration` being used for s3. But what is the problem with this then:
```
template <class Storage, class ...Args>
explicit IStorageDataLake(Storage::Configuration && configuration, Args && ...storage_args)
 : Storage(getAdjustedConfiguration(), std::forward<Args>(args)...) {}
```

[the remark this answers is not in the document]
Comment (ucasfl) [reply src/Storages/StorageS3.h]: We need first `updateConfiguration` then `getAdjustedConfiguration`...

File: src/Storages/IStorageDataLake.h [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+        String new_uri = std::filesystem::path(configuration.url.uri.toString()) / Name::data_directory_prefix
+            / MetadataParser::generateQueryFromKeys(keys, configuration.format);
+
+        typename Storage::Configuration new_configuration(configuration);
+
+        /// The only S3 related stuff remain: S3::URI
+        new_configuration.url = S3::URI(new_uri);
+
+        LOG_DEBUG(log, "Table path: {}, new uri: {}", configuration.url.key, new_uri);
+
```
Comment (ucasfl) [inline src/Storages/IStorageDataLake.h:75]: The only S3 related stuff remained is here.

Comment (kssenii) [reply src/Storages/IStorageDataLake.h:80]: we could add a method to `Storage::Configuration`:
```
configuration.appendToPath(Name::data_directory_prefix / MetadataParser::generateQueryFromKeys(keys, configuration.format))
```
and
```
class StorageS3
{
    struct Configuration
    {
         S3::URI url;
         ....
    
         void appendToPath(const std::string & suffix)
         {
               url = std::filesystem::path(url.uri.toString()) / suffix;
         }
         
         std::string getPath() const { return url.uri.toString(); } /// For logging
    }
}
```

File: src/Storages/IStorageDataLake.h [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+    static typename Storage::Configuration
+    getAdjustedConfiguration(const ContextPtr & context, const typename Storage::Configuration & configuration, Poco::Logger * log)
+    {
+        MetadataParser parser{configuration, context};
+
+        auto keys = parser.getFiles();
+        String new_uri = std::filesystem::path(configuration.url.uri.toString()) / Name::data_directory_prefix
+            / MetadataParser::generateQueryFromKeys(keys, configuration.format);
+
+        typename Storage::Configuration new_configuration(configuration);
```
Comment (kssenii) [inline src/Storages/IStorageDataLake.h:74]: you already have `using Configuration = typename Storage::Configuration` in this class
```suggestion
        Configuration new_configuration(configuration);
```
and in other places in this file

File: src/Storages/StorageS3.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (kssenii) [reply src/Storages/StorageS3.h]: This is manageable the same way too.

File: src/Storages/StorageS3.cpp [hunk 24]
```diff
@@ -1217,6 +1217,18 @@ void StorageS3::truncate(const ASTPtr & /* query */, const StorageMetadataPtr &,
...
+StorageS3::Configuration StorageS3::updateConfiguration(ContextPtr local_context, const StorageS3::Configuration & configuration)
+{
+    StorageS3::Configuration new_configuration(configuration);
+    updateS3Configuration(local_context, new_configuration);
+    return new_configuration;
+}
+
+void StorageS3::updateConfiguration(ContextPtr local_context, StorageS3::Configuration & configuration)
+{
+    updateS3Configuration(local_context, configuration);
```
Comment (kssenii) [inline src/Storages/StorageS3.cpp:1229]: why not just rename updateS3Configuration -> updateConfiguration ?

Comment (ucasfl) [reply src/Storages/StorageS3.cpp:1229]: OK

**Review by kssenii — APPROVED** [review]

**Comment by nickitat** [comment]
build without Avro is broken. let's fix or revert

**Comment by ucasfl** [comment]
> build without Avro is broken. let's fix or revert

It's because the metadata of Iceberg is Avro format, so I need Avro to parse the metadata.

**Comment by azat** [comment]
I've also hit by this, @ucasfl any PR?

**Comment by Samrose-Ahmed** [comment]
Excuse my ignorance but why does the code do listFiles on S3, is that just a scrappy V0 impl, Iceberg shouldn't require any listing, just following the metadata files.

**Comment by ucasfl** [comment]
> Excuse my ignorance but why does the code do listFiles on S3, is that just a scrappy V0 impl, Iceberg shouldn't require any listing, just following the metadata files.

Because we don't have Iceberg C++ API.

**Comment by ucasfl** [comment]
> build without Avro is broken. let's fix or revert

@nickitat Fixed in https://github.com/ClickHouse/ClickHouse/pull/46554.

**Comment by rice668** [comment]
By the way, do we also consider supporting Iceberg tables stored in HDFS ?

**Comment by kssenii** [comment]
yes
